### PR TITLE
Add compute lab

### PR DESCRIPTION
**Pull request by teodutu**

### CODE REVIEW

File: content/chapters/compute/lab/README.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+The reason is that the time required to create a process is longer than that required to create a thread.
+Because a process needs a separate virtual address space (VAS) and needs to duplicate some internal structures such as the file descriptor table and page table, it takes the operating system more time to create it than to create a thread.
+On the other hand, threads belonging to the same process share the same VAS and, implicitly, the same OS-internal structures.
+Therefore, they are more lightweight than processes.
+
+#### Libraries for Parallel Processing
+
+Our previous implementation spawned the worker threads "manually" by using the `spawn` function.
+This is **not** a syscall, but a wrapper over the most common thread-management API in Linuxs: POSIX Threads or `pthreads`.
```
Comment (razvand) [inline content/chapters/compute/lab/README.md:149]:
```suggestion
This is **not** a syscall, but a wrapper over the most common thread-management API in POSIX-based operating systems (such as Linux, FreeBSD, macOS): POSIX Threads or `pthreads`.
```

File: content/chapters/compute/lab/README.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+clone(child_stack=0x7f60b4e472b0, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tid=[1819694], tls=0x7f60b4e48640, child_tidptr=0x7f60b4e48910) = 1819694
+
+student@os:~/.../lab/support/sum-array/d/$ strace -e clone ./sum_array_processes 2
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7a4e346650) = 1820599
+clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f7a4e346650) = 1820600
+```
+
+We ran each program with an argument of 2, so we have 2 calls to `clone`.
+Notice that in the case of threads, then `clone` syscall receives more arguments.
+If we check [`clone`'s man page](https://man.archlinux.org/man/clone3.2.en) `flags` it receives specify the following:
```
Comment (razvand) [inline content/chapters/compute/lab/README.md:194]:
```suggestion
Relevant flags passed as arguments when creating threads are documented in [`clone`'s man page](https://man.archlinux.org/man/clone3.2.en):
```

File: content/chapters/compute/lab/README.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+- `CLONE_VM`: the child and the parent process share the same VAS
+- `CLONE_{FS,FILES,SIGHAND}`: the new thread shares the filesystem information, file and signal handlers with the one that created it
+The syscall also receives valid pointers to the new thread's stack and TLS, i.e. the only parts of the VAS that are distinct between threads (although they are technically accessible from all threads).
```
Comment (razvand) [inline content/chapters/compute/lab/README.md:197]: This is a rather out-of-nowhere reference to TLS. I assume it will be detailed sometime in the lab text.

File: content/chapters/compute/lab/README.md [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+By contrast, when creating a new process the `clone` syscall receives far less information.
+Remember that in both cases `clone` creates a new **thread**.
+When creating a process, `clone` creates this new thread within a separate process.
```
Comment (razvand) [inline content/chapters/compute/lab/README.md:201]:
```suggestion
When creating a process, `clone` creates this new thread within a new separate address space.
The creation of a new address space and a new thread equates to the creation of a new process.
```

**Review by razvand — CHANGES_REQUESTED** [review]
Thanks, @teodutu . This is a good start.

I think at times you go into much depth and on not-so-relevant tracks. I would keep a "main, simpler" narrative and leave the rest, such as D threading model, benchmarking internals as "extra" / bonus topics.

See my comments below.

**Review by Alex-deVis — COMMENTED** [review]
We could develop the section about creating processes by providing a `fork` + `exec` example. It would also give better insight for examples with strace where `execve` is featured.

File: content/chapters/compute/lab/README.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
+# Compute
+
+The main criterion we use to rank CPUs is their _computation power_, i.e. their ability to crunch numbers and do math.
+Numerous benchmarks exist out there and they are publicly displayed on websites such as [CPUBenchmark](https://www.cpubenchmark.net/).
+
+This benchmark measures the performance of the computer's CPU in a variety of scenarios:
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:6]: It explains what is being measured as part of a benchmark to get a better idea about what a CPU does adn why there is a need for an OS. All in all, it's less than 15 lines and then moves on introduce the OS. I don't think it's so elaborate, especially since the links aren't menat to be analysed in detail, but merely skimmed over.

File: content/chapters/compute/lab/README.md (hunk 3 above)
Comment (teodutu) [reply content/chapters/compute/lab/README.md:197]: Yes. I think this is alright since this section comes at the end of the "regular" lab.

File: content/chapters/compute/lab/README.md [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+You most likely did get some speedup, especially when using 8 processes.
+Now we will try to improve this speedup by using **threads** instead.
+
+### Threads
+
+Compile the code in `sum_array_threads.d` and run it using 1, 2, 4 and 8 threads as you did before.
+Each thread runs the `calculateArrayPartSum` function and then finishes.
+Running times should be _slightly_ smaller than the implementation using processes.
+
+The reason is that the time required to create a process is longer than that required to create a thread.
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:141]: I said above that the difference should be small. Should I remove this paragraph or rephrase it?

File: content/chapters/compute/lab/support/sum-array/d/generate_random_array.d [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module generate_random_array;
+
+
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/support/sum-array/d/generate_random_array.d:3]: To better separate differents "parts" of the code, but now that you asked, they're pretty pointless. I removed the extra line.

File: content/chapters/compute/lab/quiz/processes-speedup.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+- Because the array is split into unequal parts
++ Because of the overhead introduced by the creation of the additional processes
+- Because the algorithm is incorrect
+- Because the operating systems runs all processes sequentially on the same core
+
+## Feedback
+
+Creating a new process involves an inherent overhead.
+You have to call the loader, launch the new process, then wait for it to finish, get its return value etc.
+All this work has to be done **sequentially** by the main thread.
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/quiz/processes-speedup.md:18]: I meant to send the message that there will always be parts of an app that cannot be paralellised and, thus, the speedup will never be equal to the number of processes. But you're right that it's ambiguous. It should be clearer now.

**Comment by teodutu** [comment]
@Alex-deVis 
> We could develop the section about creating processes by providing a fork + exec example

I deliberately tried to avoid this. `fork` + `exec` is too complex to be placed at the beginning of the lab. It will be covered, but towards the end of the lab or maybe in the `Arena`, because it's an implementation detail on POSIX systems.

File: content/chapters/compute/lab/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/compute/lab/README.md:6]: OK, you can keep it. My concern is that it focuses on benchmarking and results which may be too much for students at this point to grasp.

File: content/chapters/compute/lab/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (razvand) [reply content/chapters/compute/lab/README.md:193]: I see. I would remove the blank line between them. Also, I would rephrase it to "This slight time difference is caused by process creation actions, which are costlier than thread creation actions." Or something in that note.

File: content/chapters/compute/lab/README.md [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+Use whatever API is provided by your language of choice for creating and waiting for processes.
+
+### The GIL
+
+Throughout this lab you might have noticed that there were no thread exercises _in Python_.
+If you did, you probably wondered why.
+It's not because Python does not support threads, because it does, but because of a mechanism called the **Global Interpreter Lock**, or GIL.
+As its name suggests, this is a lock implemented inside most commonly used Python interpreter (CPython), which only allows **one** thread to run at a given time.
+As a consequence, multithreaded programs written in Python run **concurrently**, not in parallel.
+For this reason, you will see no speedup even when you run an embarrassingly parallel code in parallel.
```
Comment (molecula2788) [inline content/chapters/compute/lab/README.md:679]: It should be mentioned that python threads are still useful when you have multiple io-bound tasks. Otherwise one would believe that they are completely useless.

File: content/chapters/compute/lab/README.md [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+
+Let's assume we only have one process on our system and that process knows how to add the numbers in an array.
+It can use however many resources it wants since there is no other process to contest it.
+It would probabily look like the code in `support/sum-array/d/sum_array_sequential.d`.
+The program also measures the time spent computing the sum.
+Let's compile and run it:
+
+```
+student@os:~/.../lab/support/sum-array/d/$
+
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:101]: No, it was only a duplicate line

File: content/chapters/compute/lab/README.md [hunk 11]
```diff
@@ -0,0 +1,685 @@
...
+As you can see, waiting is a very coarse form of synchronisation.
+If we only use waiting, we can expect no speedup as a result of parallelism, because one thread must finish completely before another can continue.
+We will discuss more fine-grained synchronisation mechanisms [later in this lab](#synchronisation).
+
+Also, at this point, you might be wondering why this exercise is written in D, while [the same exercise, but with processes](#practice-wait-for-me) was written in Python.
+There is a very good reason for this and has to do with how threads are synchronized by default in Python.
+You can find out what this is about [in the Arena section](#the-gil), after you have completed the [Synchronisation section](#synchronisation).
+
+#### Practice: `fork()`
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:230]: It is a problem. I moved it after `Practice: Wait for Me!`.

File: content/chapters/compute/lab/README.md [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+> The VAS of the child process initially points to the same PAS as that of the parent.
+> A (physical) frame is only duplicated by the child when it attempts to **write** data to it.
+
+This ensures that read-only sections remain shared, while writable sections are shared as long as their contents remain unchanged.
+When changes happen, the process making the change receives a unique frame as a modified copy of the original frame _on demand_.
+
+In the image below we have the state of the child and parent processes right after `fork()` returns in both of them.
+See how each has its own VAS, both of them being mapped to (mostly) the same PAS.
+
+![Copy-on-Write](../lecture/media/copy-on-write-initial.svg)
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:369]: It is present in the [lecture PR](https://github.com/systems-cs-pub-ro/operating-systems-oer/pull/5). Once that is merged, it should render properly. There's no point in duplicating it.

File: content/chapters/compute/lab/README.md [hunk 13]
```diff
@@ -0,0 +1,685 @@
...
+One uses threads while the other uses processes.
+
+Run both programs.
+The one using threads will most likely print a negative sum, while the other displays the correct sum.
+This happens because all threads refer the same memory for the array `arr`.
+What happens to the processes is a bit more complicated.
+[Later in this lab](#copy-on-write) we will see that initially, the page tables of all processes point to the same physical frames or `arr`.
+When the malicious process tries to corrupt this array by **writing data to it**, the OS duplicates the original frames of `arr` so that the malicious process writes the corrupted values to these new frames, while leaving the original ones untouched.
+This mechanism is called **Copy-on-Write** and is an OS optimisation so that memory is shared between the parent and the child process, until one of them attempts to write to it.
+At this point, this process receives its own separate copies of the previously shared frames.
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:310]: Right. I only left a forward reference to copy-on-write to make students curious.

File: content/chapters/compute/lab/README.md [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+
+By contrast, when creating a new process, the arguments of the `clone` syscall are simpler (i.e. fewer flags are present).
+Remember that in both cases `clone` creates a new **thread**.
+When creating a process, `clone` creates this new thread within a new separate address space.
+
+### Libraries for Parallel Processing
+
+In `support/sum-array/d/sum_array_threads.d` we spawned threads "manually" by using the `spawn` function.
+This is **not** a syscall, but a wrapper over the most common thread-management API in POSIX-based operating systems (such as Linux, FreeBSD, macOS): POSIX Threads or `pthreads`.
+Using `ltrace`, we can see that `spawn` calls `pthread_create` in order to spawn the new thread.
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:567]: The scheduler itself also creates some additional threads, the output from `ltrace` is messy at best. I've just linked the implementation of `spawn`. It should be clearer.

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c [hunk 15]
```diff
@@ -0,0 +1,141 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+// #include "utils.h"
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:9]:
```suggestion
#include "utils.h"
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+#define EXIT_CMD		5
+
+static char *verb;
+static char **args;
+static char *stdin_file;
+static char *stdout_file;
+static char *stderr_file;
+
+static char *value;
+static char *var;
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:27]:
```suggestion
```
```
mini_shell.c:27:14: error: ‘var’ defined but not used [-Werror=unused-variable]
   27 | static char *var;
      |              ^~~
mini_shell.c:26:14: error: ‘value’ defined but not used [-Werror=unused-variable]
   26 | static char *value;
      |              ^~~~~
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+	/**
+	 * TODO - Create a process to execute the command.
+	 * Use `execvp` to launch the new process.
+	 */
+	pid = fork();
+	switch (pid) {
+	case -1:
+		/* Error */
+		DIE(1, "fork");
+
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:95]:
```suggestion
		DIE(1, "fork");
		break;
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+	case 0:
+		/* Child process */
+		execvp(args[0], (char *const *) args);
+		DIE(1, "execvp");
+
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:100]:
```suggestion
		DIE(1, "execvp");
		break;
```

File: content/chapters/compute/lab/README.md [hunk 19]
```diff
@@ -0,0 +1,772 @@
...
+
+Also notice that we're not using hundreds or thousands of processes.
+Assuming our system has 8 cores, only 8 _threads_ (we'll see this later in the lab) can run at the same time.
+In general, **the maximum number of threads that can run at the same time is equal to the number of cores**.
+In our example, each process only has one thread: its main thread.
+So by consequence and by forcing the terminology (because it's the main thread of these processes that is running, not the processes themselves), we can only run in parallel a number of processes equal to at most the number of cores.
+
+#### Practice: Baby steps - Python
+
+Run the code in `support/create-process/popen.py`.
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:145]: The point of this exercise is to teach students how to use `Popen()` so they can do so themselves in the next exercise.

File: content/chapters/compute/lab/solution/race-condition/asm/race_condition_lock.S [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+	; pthread_create(&tids[0], NULL, increment_var, NULL)
+	mov rdi, rsp
+	xor rsi, rsi
+	mov rdx, increment_var
+	xor rcx, rcx
+	call pthread_create
+
+	; pthread_create(&tids[1], NULL, decrement_var, NULL)
+	lea rdi, [rsp + 8]
+	mov rdi, rsp
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/race-condition/asm/race_condition_lock.S:57]:
```suggestion
```

File: content/chapters/compute/lab/support/race-condition/asm/race_condition_lock.S [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+	; pthread_create(&tids[0], NULL, increment_var, NULL)
+	mov rdi, rsp
+	xor rsi, rsi
+	mov rdx, increment_var
+	xor rcx, rcx
+	call pthread_create
+
+	; pthread_create(&tids[1], NULL, decrement_var, NULL)
+	lea rdi, [rsp + 8]
+	mov rdi, rsp
```
Comment (molecula2788) [inline content/chapters/compute/lab/support/race-condition/asm/race_condition_lock.S:57]:
```suggestion
```

File: content/chapters/compute/lab/README.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/README.md:145]: You're right. I've added a few phrases regarding this in the lab text.

File: content/chapters/compute/lab/solution/sleepy/Makefile [hunk 22]
```diff
@@ -0,0 +1,5 @@
+BINARY = sleepy_creator
+include ../../../../../common/makefile/single.mk
+
+clean:
+	-rm -f syscalls*
```
Comment (molecula2788) [inline content/chapters/compute/lab/solution/sleepy/Makefile:5]: I think this is a leftover from previous versions?
```
$ make
Makefile:4: *** target file 'clean' has both : and :: entries.  Stop.
```
Same in `support/sleepy/Makefile`.

File: content/chapters/compute/lab/README.md [hunk 23]
```diff
@@ -870,13 +870,269 @@ This is because, upon creating a new thread, its TLS is initialised.
...
+One of the most important of their features is that they offer true parallelism.
+With KLTs, we can truly run a program on all the cores of our CPU at once.
+But we must pay a price for this: scheduling them is very complex and context switches are costly (in terms of time), especially when switching threads belonging to different processes. 
+
+By contrast, **user-level threads (ULT)** are managed by the user space.
+More of the ULTs created by a program are generally mapped to the same kernel thread.
+If a process only creates ULTs, then they will all be mapped to the single, main kernel thread of the process.
+So if we cannot run code in parallel with ULTs, then why use them?
+Well, for I/O-intensive programs (those that do lots of network calls or file operations like reads and writes - web servers are a good example), threads are expected to perform lots of blocking calls, which causes context switches.
+In such cases, user-level threads may be useful as context switches bring less overhead between user-level threads.
```
Comment (molecula2788) [inline content/chapters/compute/lab/README.md:905]: I'm not sure if this is correct? If a user-level thread does a blocking io call, none of the other user-level threads can run.

File: content/chapters/compute/lab/quiz/tcb-libult-unikraft.md [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
++ `start_routine` and `entry`
+- `id` and `name`
++ `context.uc_stack` and `stack`
+- `arguments` and `flags`
+- `has_dynamic_stack` and `detached`
++ `argument` and `arg`
+- `context` and `sched`
+- `context` and `tls`
++ `context` and `ctx`
++ `return_value` and `prv`
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/quiz/tcb-libult-unikraft.md:18]: Yeah, I didn't see what's to explain, but now I think I should explain why they are similar.

File: content/chapters/compute/lab/README.md (hunk 23 above)
Comment (teodutu) [reply content/chapters/compute/lab/README.md:905]: Yeah, you're right. I changed the use case to something more generic: "more context switches => ULTs are better".

File: content/chapters/compute/lab/support/apache2-simulator/apache2_simulator_condition.py [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+def worker(event, id):
+    msg = ''
+
+    while True:
+        with event:
+            while len(messages) == 0:
+                event.wait()
+
+            print(f'Worker {id} started handling message...')
+            sleep(randint(1, 5))
```
[the remark this answers is not in the document]
Comment (teodutu) [reply content/chapters/compute/lab/support/apache2-simulator/apache2_simulator_condition.py:17]: I know, but being Python threads, any form of waiting, either by sleeping or by busy-waiting will cause the same behaviour. Do you see a better approach?

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md [hunk 26]
```diff
@@ -0,0 +1,12 @@
...
+## Question Text
+
+What causes the page faults that occur between the first and second steps?
+
+## Question Answers
+
+- Calling `fork()` duplicates the pages previously allocated by the parent
++ Demand paging makes the pages in the `p` array to be mapped to frames only when written.
+- The OS duplicates the parent's pages in preparation for `fork()`
+- `mmap()` sets the pages to be mapped at a later time, decided by the OS
```
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]:
```suggestion
- `mmap()` sets the pages to be mapped at a later time, decided by the OS

## Feedback

```
Add feedback for this quiz.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+
+What causes the page faults registered by the child after the fifth step?
+
+## Question Answers
+
++ The child writes data to the frames it previously shared with its parent and the copy-on-write mechanism copies and remaps them before writing said data
+- Demand paging propagates the lazy allocation of pages from the parent to the child
+- Creating the child process inherently duplicates some frames
+- They are caused by the loader forking itself when creating the child process
+- They are caused by the `bash` process forking itself when creating the child process
```
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/child-faults-after-write.md:13]:
```suggestion
- They are caused by the `bash` process forking itself when creating the child process

## Feedback

```
Add feedback for this quiz.

File: content/chapters/compute/lab/quiz/tls-synchronization.md [hunk 28]
```diff
@@ -0,0 +1,13 @@
+# TLS Synchronisation
+
+## Question Text
+
+Is placing `var` from `support/race-condition/c/race_condition_tls.c` in the TLS a valid form of synchronisation?
+
+## Question Answers
+
++ No, because
```
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/tls-synchronization.md:9]: This is incomplete.

**Review by Alex-deVis — CHANGES_REQUESTED** [review]
I left targeted suggestions in this PR[1] to your branch.
In a broader view:
- We should decide on using `synchronisation` or `synchronization` for consistency
- We should consider naming quizzes instead of listing them as "quiz"

[1]: https://github.com/teodutu/operating-systems-oer/pull/3

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md (hunk 26 above)
Comment (teodutu) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: I wanted to, but I don't know what to write. The feedback is kind of the answer itself.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md (hunk 27 above)
Comment (teodutu) [reply content/chapters/compute/lab/quiz/child-faults-after-write.md:13]: Ditto. The question asks for the cause of the page faults and the answer also provides an explanation.

**Comment by teodutu** [comment]
> We should decide on using synchronisation or synchronization for consistency

Yeah, you're right. I switched to the American spelling :(

> We should consider naming quizzes instead of listing them as "quiz"

So you mean something like `[Quiz - Some Name Here](../quiz/blah-blah.md)`? I don't see a problem with simply writing "Quiz". In the future, they will be interactive so the name of the link is irrelevant.

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md (hunk 26 above)
Comment (Alex-deVis) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: We could leave a link to a section where demand paging is explained. If that seems ugly, no feedback is also ok.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md (hunk 27 above)
Comment (Alex-deVis) [reply content/chapters/compute/lab/quiz/child-faults-after-write.md:13]: Agreed.

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md (hunk 26 above)
Comment (teodutu) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: It's ugly and a bit unnecessary. Reading the lab section on demand paging is kind of the default feedback. I'd leave this question without feedback.

**Comment by Alex-deVis** [comment]
> So you mean something like [Quiz - Some Name Here](../quiz/blah-blah.md)?

Yes, I think it makes it easier to navigate the lab.

**Comment by teodutu** [comment]
> > So you mean something like [Quiz - Some Name Here](../quiz/blah-blah.md)?
> 
> Yes, I think it makes it easier to navigate the lab.

The other labs don't follow this convention and I don't see how it improves navigability. You can still click a link to a quiz either way.
